borg/converter.py

```diff
@@ -0,0 +1,150 @@
+        segments = [ filename for i, filename in self.io.segment_iterator() ]
+        try:
+            keyfile = self.find_attic_keyfile()
+        except KeyfileNotFoundError:
+            print("no key file found for repository")
+        else:
+            self.convert_keyfiles(keyfile, dryrun)
+        self.close()
+        self.convert_segments(segments, dryrun)
+        self.convert_cache(dryrun)
+
+    @staticmethod
+    def convert_segments(segments, dryrun):
+        '''convert repository segments from attic to borg
+
+        replacement pattern is `s/ATTICSEG/BORG_SEG/` in files in
+        `$ATTIC_REPO/data/**`.
+
+        luckily the segment length didn't change so we can just
+        replace the 8 first bytes of all regular files in there.'''
+        for filename in segments:
+            print("converting segment %s in place" % filename)
+            if dryrun:
+                continue
+            with open(filename, 'r+b') as segment:
+                segment.seek(0)
+                segment.write(MAGIC)
+
+    def find_attic_keyfile(self):
+        '''find the attic keyfiles
+
+        the keyfiles are loaded by `KeyfileKey.find_key_file()`. that
+        finds the keys with the right identifier for the repo
+
+        this is expected to look into $HOME/.attic/keys or
+        $ATTIC_KEYS_DIR for key files matching the given Borg
+        repository.
+
+        it is expected to raise an exception (KeyfileNotFoundError) if
+        no key is found. whether that exception is from Borg or Attic
+        is unclear.
+
+        this is split in a separate function in case we want to use
+        the attic code here directly, instead of our local
+        implementation.'''
+        return AtticKeyfileKey.find_key_file(self)
+
+    @staticmethod
+    def convert_keyfiles(keyfile, dryrun):
+
+        '''convert key files from attic to borg
+
+        replacement pattern is `s/ATTIC KEY/BORG_KEY/` in
+        `get_keys_dir()`, that is `$ATTIC_KEYS_DIR` or
+        `$HOME/.attic/keys`, and moved to `$BORG_KEYS_DIR` or
+        `$HOME/.borg/keys`.
+
+        no need to decrypt to convert. we need to rewrite the whole
+        key file because magic number length changed, but that's not a
+        problem because the keyfiles are small (compared to, say,
+        all the segments).'''
+        print("converting keyfile %s" % keyfile)
+        with open(keyfile, 'r') as f:
+            data = f.read()
+        data = data.replace(AtticKeyfileKey.FILE_ID,
+                            KeyfileKey.FILE_ID,
+                            1)
+        keyfile = os.path.join(get_keys_dir(),
+                               os.path.basename(keyfile))
+        print("writing borg keyfile to %s" % keyfile)
+        if not dryrun:
+            with open(keyfile, 'w') as f:
+                f.write(data)
+            with open(keyfile, 'r') as f:
+                data = f.read()
+            assert data.startswith(KeyfileKey.FILE_ID)
```

Comment:
i think that's a remnant of the code's past life as a unit test... it's now a dupe of the code in the unit tests, so removed, indeed.

i indeed didn't trust `f.write()`, i had a bunch of problems with the inplace edits - turns out that mode `a` can't `seek()`...